aws/credentials/process_provider.go

```diff
@@ -0,0 +1,228 @@
+	// for testing
+	executionFunc func(process string) ([]byte, error)
+}
+
+// NewProcessProvider returns a pointer to a new Credentials object
+// wrapping the Profile file provider.
+func NewProcessProvider(filename, profile string) *Credentials {
+	return NewCredentials(&ProcessProvider{
+		Filename:      filename,
+		Profile:       profile,
+		executionFunc: executeCredentialProcess,
+	})
+}
+
+// Retrieve executes the 'credential_process' and returns the credentials
+func (p *ProcessProvider) Retrieve() (Value, error) {
+	filename, err := p.filename()
+	if err != nil {
+		return Value{ProviderName: ProcessProviderName}, err
+	}
+
+	creds, err := p.loadProfile(filename, p.profile())
+	if err != nil {
+		return Value{ProviderName: ProcessProviderName}, err
+	}
+
+	return creds, nil
+}
+
+// loadProfiles executes the 'credential_process' from the profile of the given
+// file, and returns the credentials retrieved. An error will be returned if it
+// fails to read from the file, or the data is invalid.
+func (p *ProcessProvider) loadProfile(filename, profile string) (Value, error) {
+	config, err := ini.Load(filename)
+	if err != nil {
+		return Value{ProviderName: ProcessProviderName}, awserr.New("ProcessProviderLoad", "failed to load shared config file", err)
+	}
+
+	var iniProfile *ini.Section
+	if len(profile) == 0 || profile == "default" {
+		iniProfile, err = config.GetSection("default")
+	} else {
+		iniProfile, err = config.GetSection("profile " + profile)
+	}
+	if err != nil {
+		return Value{ProviderName: ProcessProviderName}, awserr.New("ProcessProviderLoad", "failed to get profile", err)
+	}
+
+	process, err := iniProfile.GetKey("credential_process")
+	if err != nil {
+		return Value{ProviderName: ProcessProviderName}, awserr.New("ProcessProviderLoad", "No credential_process specified", err)
+	}
+	return p.credentialProcess(process.String())
+}
+
+type credentialProcessResponse struct {
+	Version         int
+	AccessKeyID     string `json:"AccessKeyId"`
+	SecretAccessKey string
+	SessionToken    string
+	Expiration      string
+}
+
+func executeCredentialProcess(process string) ([]byte, error) {
+	processArgs := strings.Split(process, " ")
+	var command string
+	var cmdArgs []string
+
+	if len(processArgs) == 1 {
+		command = processArgs[0]
+	} else if len(processArgs) > 1 {
+		command = processArgs[0]
+		cmdArgs = processArgs[1:]
+	}
+
+	// TODO: Check for executability?
+	_, err := os.Stat(command)
+	if os.IsNotExist(err) {
+		return nil, awserr.New("ProcessProviderLoad", "credential_process "+command+" not found", err)
+	} else if err != nil {
+		return nil, awserr.New("ProcessProviderLoad", "failed to open credential_process "+command, err)
+	}
+
+	// Execute command
+	cmd := exec.Command(command, cmdArgs...)
+	cmd.Env = os.Environ()
+	out, err := cmd.Output()
+	if err != nil {
+		return nil, awserr.New("ProcessProviderLoad", "Error executing credential_process", err)
```

Comment:
Should this capture `stderr` as well. I don't remember if the `err` returned by `cmd.Output` includes stderr or not. This might only include the exit code.